(artifact-parser): Add the Version header parser

oleorhagen opened the pull request:
This adds a simple parser for the Mender-Artifact v3 version header.

As well as a small utility ByteWriter class to io.

I do have to admit that I feel this turned out needlessly complicated, so all open for advice on how to improve this!

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	const json::ExpectedJson expected_version = expected_json.value();
+
+	if (!expected_version) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	auto expected_obj = expected_version.value();
```
I'm somewhat confused by this code. Is there three layers of ExpectedJson inside each other? Line 64 looks like a no-op, you are just assigning a payload value to an ExpectedJson, and since it is not an error, it will always construct an ExpectedJson with a value, and the next `if` will always evaluate false.

But feel free to enlighten me if I'm misunderstanding this.

oleorhagen replied:
I don't even know. I think this is horrible...

However, I have not come up with a cleaner way...

oleorhagen replied:
Alright, cleaning it up a little...

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+
+	if (!expected_obj.IsObject()) {
+		return MakeError(
+			ParseError,
```
The JSON is valid though, so should this be `FormatError` instead? It's JSON, but a format we don't understand.

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	auto version = expected_obj.Get("version");
+	if (!version) {
+		return MakeError(
+			ParseError,
```
Same here, maybe `FormatError`.

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+			"Failed to unmarshal the version from the version JSON data. No 'version' key found");
+	}
+	auto version_unwrapped = version.value().GetInt();
+	int version_;
```
Trailing underscores are for member variables.

kacf commented on artifact/version/version.hpp:
```diff
@@ -0,0 +1,49 @@
+#include <common/io.hpp>
+
+namespace artifact {
+namespace v3 {
```
👍

But should the files be inside `v3` as well?

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	}
+	version_ = version_unwrapped.value();
+
+	auto format = expected_obj.Get("format");
+	if (!format) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. No 'format' field found");
+	}
+	auto format_unwrapped = format.value().GetString();
+	if (!format_unwrapped) {
+		return MakeError(
+			ParseError,
+			"Failed to unmarshal the version from the version JSON data. The 'format' fields needs to contain a string");
+	}
```
Suggestion: You could shorten this by doing:

```c++
auto format = expected_obj.Get("format");
if (!format || !format.value().GetString()) {
	return MakeError(
		ParseError,
		"Failed to unmarshal the version from the version JSON data. 'format' field not found or not a string");
}
auto format_unwrapped = format.value().GetString();
```

And similar with `version` above. Not saying it's better though, verbosity is always good. Just a suggestion.

kacf commented on artifact/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	if (version_ != 3) {
+		return MakeError(
+			VersionError,
+			"Failed to unmarshal the version from the version JSON data. Only 'version' 3 is supported received version "
```
> Failed to unmarshal the version from the version JSON data

You've used this message in nearly every error so far. Maybe you should make it part of the error code string instead, and use only the more specific strings in this function?

kacf commented on artifact/version/version.hpp:
```diff
@@ -0,0 +1,49 @@
+namespace error = mender::common::error;
+
+
+enum ErrorCode {
+	NoError = 0,
+	ParseError,
+	VersionError,
+	FormatError,
```
I have a question about this, but it's an abstract question, hard to put into words here. Ping me when you have time.

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, "Failed to parse the version JSON");
```
Might as well just return `expected_json.error()` here.

oleorhagen replied:
Can't directly actually, since they are different types...

But I can add a conversion function?

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
```
We copy the data into a buffer only to create a string from it here and throw the buffer away. It would be needed if `StringFromByteVector` took a reference and potentially modified the given buffer, but it is (I believe) a `const` reference and so we can safely pass in a buffer without creating a copy of it.

oleorhagen replied:
The absolute best would be if `json` took a `Reader`, and I wouldn't have to do any of this.

Since ur intimately familiar with this. Would it be a lot of work to add you think?

vpodzime replied:
Yeah, I think I can do that. The JSON stuff we have predates our IO stuff so it wasn't possible back then. The only complication is that to do it right, we should have a UTF8Reader because bytes shouldn't be passed into the JSON parser that expects characters.

oleorhagen replied:
Great :tada: :tada:

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	string format {"unknown"};
+
+
+	auto err = io::Copy(bw, reader);
```
No a real issue here, I believe, but what if the whole JSON doesn't fit into the buffer?

oleorhagen replied:
Then we've received some mumbo-jumbo which we don't support anyways :smile:

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+		return MakeError(FormatError, "No version field found");
+	}
+
+	const json::ExpectedInt value_version = version_value.value().GetInt();
```
Out of  scope of this PR, but `ExpectedInt` should be in the `expected` namespace/header, I think.

vpodzime replied:
Please read this line (maybe out loud) and I'm pretty sure you will be able to come up with better variable names.

oleorhagen replied:
hahaha. I think so too

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+			return MakeError(
+				VersionError,
+				"Failed to unmarshal the version from the version JSON data. Only 'version' 3 is supported received version "
+					+ std::to_string(version));
```
Failing to get the version integer should also be an error, right?

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	const json::ExpectedJson format_value = version_json.Get("format");
+	if (version_value) {
```
`if (format_value)`?

kacf replied:
Yes, plus code continues on error here as well.

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+				return MakeError(
+					FormatError,
+					"Failed to unmarshal the 'format' from the version JSON data. The client only understands the 'mender' Artifact type. Got format: "
+						+ format);
```
Again, `!value_format` is an error too, right?

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+		}
+	}
+
+	return Version {.version = 3, .format = "mender"};
```
It feels a bit strange to have a function called `Parse()` that doesn't actually return any parsed data, but not a big deal.

oleorhagen replied:
Well, it does, right. The parsed values match exactly this. I just wanted to highlight it here. But if it's odd I can change it.

vpodzime commented on artifact/v3/version/version.hpp:
```diff
@@ -0,0 +1,63 @@
+struct Version {
+	const int version;
+	const string format;
+};
```
Similar here, a bit weird to have a type for something that's actually just a hard-coded constant pair `3-"mender"`.

oleorhagen replied:
Maybe, but it does match the structure of what is parsed to a code-template of sorts. So at least it's symmetric with how we handle the rest of the parsing.

vpodzime commented on artifact/v3/version/version_test.cpp:
```diff
@@ -0,0 +1,133 @@
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+#include <fstream>
```
`sstream`?

kacf commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	const json::ExpectedInt value_version = version_value.value().GetInt();
+	if (value_version) {
+		version = value_version.value();
+		if (version != 3) {
```
Come to think of it, maybe this should be a constant.

kacf commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	const json::ExpectedInt value_version = version_value.value().GetInt();
+	if (value_version) {
```
This logic is wrong now. If the `version` field is not an int, the code continues, but it should fail.

oleorhagen replied:
Oh

My

God....

oleorhagen commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,124 @@
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
```
Hmm... Why not?

oleorhagen commented on common/io.hpp:
```diff
@@ -108,6 +109,26 @@ class StringReader : virtual public Reader {
 
+class ByteWriter : virtual public Writer {
+private:
+	vector<uint8_t> &receiver;
```
I should really add a static check for this in my editor or smthn...

oleorhagen commented on common/io.hpp:
```diff
@@ -108,6 +109,26 @@ class StringReader : virtual public Reader {
+	ExpectedSize Write(const vector<uint8_t> &dst) override {
+		int i = 0;
+		for (; i < std::min(dst.size(), receiver.size() - bytes_written_); ++i) {
+			receiver[i + bytes_written_] = dst[i];
```
We can unroll the loop? :p 

But really, I'll wrap it one of the std::copy derivatives

oleorhagen commented:
Rebased on top of: https://github.com/mendersoftware/mender/pull/1173

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+
+	const int version = version_json.Get("version")
+							.and_then([](const json::Json &j) { return j.GetInt(); })
+							.value_or(-1);
```
Is this what our formatter produces? :scream:

oleorhagen replied:
Yes :see_no_evil: Not the prettiest on the :eyes:

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+
+	const int version = version_json.Get("version")
+							.and_then([](const json::Json &j) { return j.GetInt(); })
+							.value_or(-1);
```
I'd drop the `value_or()` and got the `ExpectedInt()` from the chain because `value_or()` discards the error.

vpodzime replied:
`value_or()` is still there.

oleorhagen replied:
Same as above. You need to sleep in more :laughing:

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+	if (version != supported_parser_version) {
+		return expected::make_unexpected(MakeError(
+			FormatError,
+			"Failed to unmarshal the version from the version JSON data. Only 'version' 3 is supported in this client"));
```
Should use the exact error (message) our JSON handling has the ability to provide.

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+
+	const string format = version_json.Get("format")
+							  .and_then([](const json::Json &j) { return j.GetString(); })
+							  .value_or("unknown");
```
Same here.

vpodzime commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+			"Failed to unmarshal the format from the version JSON data. Only the 'mender' format is supported in this client"));
+	}
+
+	return Version {.version = 3, .format = "mender"};
```
Should use the constants.

oleorhagen replied:
Yeh, let's. I left it there to highlight it, but it's really just duplicating...

vpodzime reviewed (commented):
+Still not a big fan of using the expected library without even trying to expand our own code, but looks like you guys are decided.

oleorhagen commented on artifact/v3/version/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  gmock
+  common_io
+  common_error
+  common_json
```
Yeh, it's good you pointed this out, I forgot to mention it.

So I _did not_ make a module out of this, since it is so small, in which case, the test needs this to link.

However maybe this was a bad idea. I really don't know. 

Since it caught you by surprise, maybe I should make a separate library out of it, it just felt a bit extensive.

I'll let you decide :pray:

oleorhagen commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
```
Yes, true. This is actually a really good rule!

oleorhagen replied:
Ur just up too early in the morning :p :sunrise_over_mountains:  :rooster:

oleorhagen commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
```
Yes, but like we discussed previously, the very best would be `json::From(io::Reader&)`

vpodzime replied:
That will require that function to read a file into a buffer anyway. Or a stream interface to our `Reader` because C++ already has IO utilities -- the streams -- an things are setup to work with them not our Go-like implementation.

oleorhagen replied:
This is true, it is just a question about the size of the buffer. Like buffering the whole payload, or doing it in chunks.

I actually had a conversation with @kacf about it this morning, where I suggested to move to the stdlib interface, and his argument is that the `io-streams` add too much code bloat for us, as they are all templated.

I'm not quite sure I fully agree, but I do understand the argument, and it is not our biggest thorn in our sides imo.

Therefore for now, going with this intermediary buffering is the way it seems.

oleorhagen commented on artifact/v3/version/version.cpp:
```diff
@@ -0,0 +1,106 @@
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	auto err = io::Copy(bw, reader);
```
At least from what I'm used to, `Copy` is nice ,because it guarantees to read to the end. A single `Read` call can theoretically do short reads, and may need to be called again.

vpodzime replied:
But the data is shuffled around (into an internal buffer in `Copy` and then into the `ByteWriter` above) and you can easily get an error if `buf` above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

oleorhagen replied:
This is true in the general case. But the advantage here is that we control the format, and the `version` header is never going to be bigger than what is here available.

oleorhagen replied:
> But the data is shuffled around (into an internal buffer in Copy and then into the ByteWriter above) and you can easily get an error if buf above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

Isn't this what I am doing with the `byte-buffer` above already?

vpodzime replied:
Assuming it's at most 1024 bytes. But let's not ride a dead horse here anymore. I'll try to come up with something for this myself and then we can have a discussion there.

vpodzime replied:
> So there are 3 copies of the same thing, all with read-only access.

It's 4 actually. `Copy` has its own internal buffer too.

vpodzime reviewed (commented):
DId you forget to push some commits? Many of the conversations are marked as resolved without the changes in the code being done.

oleorhagen commented:
> DId you forget to push some commits? Many of the conversations are marked as resolved without the changes in the code being done.

Yes, just fixed locally, not pushed yet. I didn't know I had re-requested a review here, I'm sorry.
Fixed and pushed now @vpodzime :smile_cat:

oleorhagen commented:
I will not merge this until the `byte-writer` PR is merged!

oleorhagen commented:
> I will not merge this until the `byte-writer` PR is merged!

Merged!

vpodzime reviewed (approved):
Just a "nitpick of nitpicks". ACK!
